index.js

```diff
@@ -28,25 +20,36 @@ function init(opts) {
     sharedLogger = new loggers.BufferedMetricsLogger(opts);
 }
 
-// This is meant to be curried via bind() so we don't have
-// to write wrappers for each metric individually.
-function callOnSharedLogger(funcName) {
-    if (sharedLogger === null) {
-        init();
-    }
-    const args = Array.prototype.slice.call(arguments, 1);
-    sharedLogger[funcName].apply(sharedLogger, args);
+/**
+ * Wrap a function so that it gets called as a method of `sharedLogger`. If
+ * `sharedLogger` does not exist when the function is called, it will be
+ * created with default values.
+ * @template {Function} T
+ * @param {T} func The function to wrap.
+ * @returns {T}
+ */
+function callOnSharedLogger(func) {
+    // @ts-expect-error Can't find a good way to prove to the TypeScript
+    // compiler that this satisfies the types. :(
+    return (...args) => {
+        if (sharedLogger === null) {
+            init();
+        }
+        return func.apply(sharedLogger, args);
+    };
```

Comment:
This isn’t *strictly* necessary, but it means we don’t have to specify the types for all the top level functions (`flush`, `gauge`, `increment`, etc.) twice. Because the previous setup referenced functions as strings, TypeScript couldn’t figure out the signatures from the `BufferedMetricsLogger` class and copy them. With this setup, it can.

For example, the old style generated types like:

```ts
export declare const flush: any;
export declare const gauge: any;
export declare const increment: any;
export declare const histogram: any;
export declare const distribution: any;
export declare const BufferedMetricsLogger: typeof loggers.BufferedMetricsLogger;
export { reporters };
```

But this gets us real signatures:

```ts
export declare const flush: (onSuccess?: () => void, onError?: (error: Error) => void) => void;
export declare const gauge: (key: string, value: number, tags?: string[], timestampInMillis?: number) => void;
export declare const increment: (key: string, value?: number, tags?: string[], timestampInMillis?: number) => void;
export declare const histogram: (key: string, value: number, tags?: string[], timestampInMillis?: number, options?: any) => void;
export declare const distribution: (key: string, value: number, tags?: string[], timestampInMillis?: number) => void;
export declare const BufferedMetricsLogger: typeof loggers.BufferedMetricsLogger;
export { reporters };
```